amp-beta-tester.php

```diff
@@ -0,0 +1,204 @@
+	echo "<div class='notice notice-error'><p><strong>{$error}</strong></p></div>";
+}
+
+/**
+ * Modifies the AMP plugin manifest to point to a new beta update if one exists.
+ *
+ * @param \stdClass $updates Object containing information on plugin updates.
+ * @return \stdClass
+ */
+function update_amp_manifest( $updates ) {
+	if ( ! isset( $updates->no_update ) ) {
+		return $updates;
+	}
+
+	$amp_zip_file = 'amp.zip';
+	$amp_manifest = isset( $updates->response[ AMP_PLUGIN_FILE ] )
+		? $updates->response[ AMP_PLUGIN_FILE ]
+		: $updates->no_update[ AMP_PLUGIN_FILE ];
+
+	$github_releases = get_amp_github_releases();
+
+	if ( is_array( $github_releases ) ) {
+		$amp_version = get_plugin_data( WP_PLUGIN_DIR . '/' . AMP_PLUGIN_FILE )['Version'];
+		$amp_updated = false;
+
+		foreach ( $github_releases as $release ) {
+			if ( $release->prerelease ) {
+				$release_version = $release->tag_name;
+
+				// If there is a new release, let's see if there is a zip available for download.
+				if ( version_compare( $release_version, $amp_version, '>' ) ) {
+					foreach ( $release->assets as $asset ) {
+						if ( $amp_zip_file === $asset->name ) {
+							$amp_manifest->new_version = $release_version;
+							$amp_manifest->package     = $asset->browser_download_url;
+							$amp_manifest->url         = $release->html_url;
+
+							// Set the AMP plugin to be updated.
+							$updates->response[ AMP_PLUGIN_FILE ] = $amp_manifest;
+							unset( $updates->no_update[ AMP_PLUGIN_FILE ] );
+
+							$amp_updated = true;
+							break;
+						}
+					}
+
+					if ( $amp_updated ) {
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	return $updates;
+}
+
+/**
+ * Fetch AMP releases from GitHub.
+ *
+ * @return array|null
+ */
+function get_amp_github_releases() {
+	$raw_response = wp_remote_get( 'https://api.github.com/repos/ampproject/amp-wp/releases' );
+	if ( is_wp_error( $raw_response ) ) {
+		return null;
+	}
+	return json_decode( $raw_response['body'] );
+}
+
+/**
+ * Replace the 'View version details' link with the link to the release on GitHub.
+ *
+ * @param string $file Plugin file.
+ * @param array  $plugin_data Plugin data.
+ */
+function replace_view_version_details_link( $file, $plugin_data ) {
+	$plugin_version = $plugin_data['Version'];
+
+	if ( is_pre_release( $plugin_version ) ) {
+		ob_start();
+		?>
+		<script>
+			document.addEventListener('DOMContentLoaded', function() {
+				const link = document.querySelectorAll("[data-slug='amp'] a.thickbox.open-plugin-details-modal");
+
+				link.forEach( (link) => {
+					link.className = 'overridden'; // Override class so that onclick listeners are disabled.
+					link.target = '_blank';
+					link.href = '<?php echo $plugin_data['url']; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>';
+				} );
+			}, false);
+		</script>
+		<?php
+
+		echo ob_get_clean(); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```

Comment:
Output buffering doesn't really make sense here. I'd just do:

```php
function replace_view_version_details_link( $file, $plugin_data ) {
	$plugin_version = $plugin_data['Version'];
	if ( ! is_pre_release( $plugin_version ) ) {
		return;
	}
	?>
		<script>...</script>
	<?php
}
```

Reply:
> > Upgrading to `1.4.0-beta1` is not recommended, since it does not include this plugin
> 
> When I am running the current beta / `develop` branch, does it even suggest upgrading to `1.4.0-beta1`? If so, that sounds wrong.

@swissspidy, fixed.